### PR TITLE
Add sapling nullifier set

### PR DESCRIPTION
**Pull request by Eirik0**
PR for https://github.com/zcash/zcash/issues/3057

### CODE REVIEW

**Comment by Eirik0** [comment]
There are a few places where I added comments like "// TODO ... sapling nullifiers" which will need to be addressed once sapling nullifiers are implemented.

**Comment by bitcartel** [comment]
@Eirik0 In a meeting @str4d @daira @bitcartel agreed that it would be a good design to prepare for future nullifier sets, beyond sapling.  So use of `bool isSapling` could be replaced with an enum identifier.  

One area to investigate then would be how to avoid a collision / avoid exhausting the single char written to the database, which is used to distinguish records on reading.  In this case, distinguishing between many nullifier sets including `DB_NULLIFIER = 's'; DB_SAPLING_NULLIFIER = 'S';`

**Comment by ebfull** [comment]
Do we want this PR to rebase on top of the v4 transaction format changes, and fix the TODOs? I think so. I'm doing the same thing with my IncrementalMerkleTree PR.

**Comment by daira** [comment]
@bitcartel wrote:
> One area to investigate then would be how to avoid a collision / avoid exhausting the single char written to the database, which is used to distinguish records on reading. In this case, distinguishing between many nullifier sets including ``DB_NULLIFIER = 's'; DB_SAPLING_NULLIFIER = 'S';``

We can cross the single-char-exhaustion bridge when we come to it, or when we change the database format (which we plan to do anyway). In the meantime we just need a treestate enum, that is mapped to a char manually chosen for each treestate when writing to the db.

**Comment by Eirik0** [comment]
@bitcartel I changed bool flag "isSapling" to an enum.

File: src/txmempool.cpp [hunk 1]
```diff
@@ -248,12 +252,18 @@ void CTxMemPool::removeConflicts(const CTransaction &tx, std::list<CTransaction>
             if (it != mapNullifiers.end()) {
                 const CTransaction &txConflict = *it->second;
                 if (txConflict != tx)
-                {
                     remove(txConflict, removed, true);
-                }
```
Comment (str4d) [inline src/txmempool.cpp]: Leave these brackets as-is.

File: src/txmempool.cpp [hunk 2]
```diff
@@ -118,7 +121,6 @@ bool CTxMemPool::addUnchecked(const uint256& hash, const CTxMemPoolEntry &entry,
     return true;
 }
 
-
```
Comment (str4d) [inline src/txmempool.cpp]: Leave this whitespace in (unnecessary part of the diff).

**Comment by str4d** [comment]
@zkbot try

**Comment by Eirik0** [comment]
@str4d The changes from the review have been made and pushed.

**Comment by str4d** [comment]
@zkbot r+

**Comment by str4d** [comment]
All transient failures.

@zkbot retry

**Pull request merged**